### PR TITLE
fix a bug in credential parsing at Job Template launch time

### PR DESCRIPTION
**Pull request by ryanpetrello**
the "merge launch credentials with what's on the JT" behavior should
_only_ be applied to jobs launched from a workflow; explicit credentials
provided in a vanilla Job Template launch payload should be used
explictly

see: https://github.com/ansible/awx/issues/932
related: https://github.com/ansible/awx/pull/740

### CODE REVIEW

**Comment by ryanpetrello** [comment]
Is there a github tag for `state:this fixes it probably, but let @alancoding tell me the _correct_ way to do this`?

This change _definitely_ resolves https://github.com/ansible/awx/issues/932, and it passes tests, but I'm a little fuzzy on where we landed on some of the actual requirements.  One thing I'm unclear on is whether this behavior was intentional on @AlanCoding's part; I suspect it could be given that https://github.com/ansible/awx/pull/740/commits/34a8e0a9b68823a1fe92fc6cd5184b0e6fb5f39c includes this snippet:

> change modern manual launch to use merge behavior

@wenottingham I know you and @AlanCoding discussed this at length.  Do you know where we landed?

I don't think @mabashian is blocked on the bug (he's said he's able to work around it for now).  At the very least, if he wants, he can develop the UI bits against this PR to confirm that the frontend is working as anticipated.

So no rush on merging this PR - this portion of the code is fairly complicated, and I'd _really_ like to pick @AlanCoding's brain a bit before merging it.

File: awx/api/views.py
Comment (ryanpetrello) [inline awx/api/views.py:2879]: Support for merging the legacy `credential` and `vault_credential` `/api/v1/` launch keys _was_ being accomplished as a side effect by this block:

https://github.com/ansible/awx/pull/973/files#diff-160977429605d785dda624eb08043a88L381

**Comment by ryanpetrello** [comment]
Disclaimer: the change in this PR is potentially controversial, and currently undergoing discussion at: https://github.com/ansible/awx/issues/932

**Review by matburt — APPROVED** [review]

**Comment by shanemcd** [comment]
Looks like this has been approved. Can we merge it?

**Comment by FloThinksPi** [comment]
This would probably also fix #688

**Comment by ryanpetrello** [comment]
@shanemcd please don't for now; I want to discuss this change in-depth with @wenottingham, @mabashian, and other maintainers when @AlanCoding returns.

**Comment by FloThinksPi** [comment]
I merged this PR in 1.0.3.0 and it seems as this does not fix #688 

image

**Comment by AlanCoding** [comment]
There is kind of an invisible house of cards that will start collapsing with this change. It follows a set of steps along the lines of:

 - Launch a JT, modifying the credentials list on launch
 - Modify the JT credential list
 - Relaunch the previous job

The expected results of these events is not at all obvious to me. This is largely the reason I pushed hard for "use merge logic everywhere". That's the only way I can set a consistent expectation. If everything is "merge credentials" everywhere, then the sequence of events has an obvious outcome - the _additional_ credentials given on the first launch are merged with the _current_ set of JT credentials.

Variants on this sequence of events will come along the lines of:

 - Give a user admin role to a credential
 - As that user, launch a JT, providing the credential
 - Revoke the user's access to that credential
 - As that user, relaunch the previous job

The expectation here is fairly clear - the user should be denied the launch. Relaunching would involve use of the prompted credential, which the user does not have.

Another variant is to

 - Launch a JT, providing a credential
 - Create a schedule from the resultant job
 - Modify the credentials that the JT uses

This would leave a schedule running that will create jobs with credentials for which we do not have a clear expectation for. Again, we do have a clear expectation with the "merge credentials everywhere" principle, which is being removed in this PR.

Of course, the feature to schedule a job is new, so it can certainly be removed. But generally, we have this blob of prompts associated with a launch, and that would lose meaning altogether. Returning to the case of relaunching without permission to the previously-provided credential, we would be left with either false positives or false negatives (I don't know which without digging).

For me, what really matters most is how the credential selector is implemented in the UI in #1096, because if the UX does not reflect what we're actually doing, we'll have violations of expectations one way or the other. The UI credentials selector will have multiple types of constraints enforced. I don't yet know what that looks like, but that's the direction I will look to start digging in.

**Pull request closed without merging**
